odoo/osv/expression.py

```diff
@@ -943,6 +943,9 @@ def _get_expression(comodel, left, right, operator):
 
                 elif field.translate is True and right:
                     need_wildcard = operator in ('like', 'ilike', 'not like', 'not ilike')
+                    if operator.endswith('like') and isinstance(right, str) and (len(right) - len(right.rstrip('\\'))) % 2:
```

Comment:
> But it returns names which don't contain \ even though the user literally asked for that. If the user searches on foo\ they should get stuff which contain `foo\`.

Now we allow users to use wildcard characters `%`, `_` and escape `\` character in the searchbar. And if they want to search stuff which contain `foo\`, they should type `foo\\`
In PostgresQL, `"name" ILIKE "%foo\bar%"` is equal to `"name" ILIKE "%foobar%"`, so I think droping meaningless `\` is OK

> I don't think many (if any) people think of the domain-level [i]like as having pattern semantics.

I would say as a developer if the [i]like for domain support pattern semantics, he should know. But for client users, I agree they should know nothing about the pattern semantics.
If we can do everything from scratch, I think 
1. pattern semantics should be supported by domain because it is useful for developers.
2. developers should wrap content with `%` explicitly themselves, as a result the current `ilike` should work like `=ilike`. And the web client should decide what to send to the backend, escaping everything to forbid the wildcard or send the raw text to the bankend as a pattern or wrap the content with `%`.

But it will break so many codes.

Usually, `\` won't appear in the business content, users use it
1. they want to type `Enter` but enter `\` by accident
2. they want to type `\%` or `\\` or `\_` but webclient send the rpc before they finish the typing

Checking it from either frontend or backend is ok for me.

Reply:
The issue we have here is that the current situation and common sense don't match.

Common sense would prescribe something like:
* operators `like`, `ilike`, `not like` and `not ilike` are meant for "contains", and _should_ escape wildcards;
* operators `=like` and `=ilike` are meant to support wildcards.

However, the current situation is such that wildcards are _de facto_ supported in the first case, because they are not escaped.  But I believe that this is an implementation gap instead of an explicit design.  The problem is that if we remove that feature, it is going to be perceived as a regression. @bouvyd